src/verilator/verilator_object.cc

```diff
@@ -0,0 +1,260 @@
+#include "base/logging.hh"
+#include "sim/sim_exit.hh"
+#include "verilator_object.hh"
+
+void
+VerilatorCPUMemPort::sendPacket(PacketPtr pkt)
+{
+    panic_if(blockedPacket != nullptr, "Should never try to send if blocked!");
+    if (!sendTimingReq(pkt)) {
+        blockedPacket = pkt;
+    }
+}
+
+void
+VerilatorCPUMemPort::recvReqRetry()
+{
+    assert(blockedPacket != nullptr);
+
+    PacketPtr pkt = blockedPacket;
+    blockedPacket = nullptr;
+
+    sendPacket(pkt);
+}
+
+bool
+VerilatorCPUMemPort::recvTimingResp(PacketPtr pkt)
+{
+    return owner->handleResponse(pkt);
+}
+
+PacketPtr
+VerilatorObject::buildPacket(const RequestPtr &req, bool read)
+{
+    return read ? Packet::createRead(req) : Packet::createWrite(req);
+}
+
+void
+VerilatorObject::buildPayloadForWrite(RequestPtr &data_req, uint8_t * data,
+        uint8_t * &packeddata)
+{
+    //get mask and figure out if we are trying to write a byte, half word or
+    //a word
+    uint8_t mask = dut.Top__DOT__tile__DOT__memory__DOT__async_data_dw_mask;
+    bool half = mask == 0x3 || mask == 0x6 || mask == 12;
+    bool byte = mask == 0x1 || mask == 0x2 || mask == 0x4 || mask ==0x8;
+    //setup request for appropiatly sized packet
+    if (byte){
+        data_req = std::make_shared<Request>(
+                new Request(
+                dut.Top__DOT__tile__DOT__memory__DOT__async_data_dw_addr),
+                1);
+        packeddata = new uint8_t;
+    }else if ( half ){
+        data_req = std::make_shared<Request>(
+                new Request(
+                dut.Top__DOT__tile__DOT__memory__DOT__async_data_dw_addr),
+                2);
+        packeddata = new uint8_t[2];
+    }else{
+        data_req = std::make_shared<Request>(
+                new Request(
+                dut.Top__DOT__tile__DOT__memory__DOT__async_data_dw_addr),
+                4);
+        packeddata = data;
+    }
+    int j = 0;
+    for (int i = 0; i < 4; ++i){
+        //if the mask is active then use this byte in our packed data
+        if (mask & (0x1 << i)){
+                //get index of 1st occurance of a 1 for address offset
+                if (j == 0)
+                    j = i;
+                packeddata[i] = data[i];
+        }
+    }
+    //ofset address based on data size/type
+    data_req->setPaddr(data_req->getPaddr() + j);
+}
+
+void
+VerilatorObject::sendData(const RequestPtr &req, uint8_t *data, bool read)
+{
+    PacketPtr pkt = buildPacket(req, read);
+    pkt->setData(data);
+    pkt->allocate();
```

Comment:
I don't think you want both setData and allocate. You want one or the other.

Also, when is data deleted? (This is why it might be easier to use allocate() than setData.)

Reply:
```
pkt->allocate()
pkt->setData(data)
delete[] data;
```